### ISSUE TITLE
Linter should recognize and correctly handle casted object literals

### ISSUE THREAD

**Issue by Z3roCoder** [palantir/tslint#1868]
### Bug Report

- __TSLint version__: `4.0.2`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: Visual Studio Code plugin for TSLint (eg2.tslint) version 0.7.1

#### TypeScript code being linted

```ts
interface ITest {
    a: number;
}
function test() {
    return <ITest>{ a: 3, b: 3 };
}
function test2() {
    const c: ITest = { a: 3, b: 3 }
}
function test3():ITest {
    return { a: 3, b: 3 };
}
async function test4():Promise<ITest> {
    return { a: 3, b: 3 };
}
```

with `tslint.json` configuration:

```json
Not relevant (I think)
```

#### Actual behavior

TS Compiler and TSlint will let test1 pass
TS Compiler complains about test2 having too many items in the object
TS Compiler complains about test3 having too many items in the object
TS Compiler and TSlint will let test4 pass

#### Expected behavior

There should be a (optional) rule that complains about casting an object literal (like in test1).

TS Compiler will let test1 pass, but TSLint will warn the user that there might be consequences in compatibility with other endpoints.

I dont know if it is possible to correctly check for test4 as well.

**Comment by andy-hanson** [comment]
Last issue is Microsoft/TypeScript#12423

**Comment by JoshuaKGoldberg** [comment]
@Z3roCoder I'm confused. Is this not resolved by TypeScript's "Object literal may only specify known properties"?

https://stackoverflow.com/questions/31816061/why-am-i-getting-an-error-object-literal-may-only-specify-known-properties

**Comment by Z3roCoder** [comment]
No, it isn't. As far as I see it, all four Variants should have the same result. Either it passes, or the Linter throws an error. Because after all, all four functions return a object of type ITest, of which the interface as only "a" defined, but also contains the property "b"

It seems inconsistent to me

**Comment by Z3roCoder** [comment]
(Okay, Test2 doesn't 'return' it but you get the gist)

**Comment by JoshuaKGoldberg** [comment]
Still a little confused here: TypeScript does complain on cases 2, 3, and 4, and TSLint's `no-object-literal-type-assertion` complains on case 1. Using your code above and the following `tslint.json`:

```json
{
    "rules": {
        "no-object-literal-type-assertion": true
    }
}
```

...this is what I get:

```powershell
PS C:\Code\test> tsc
index.ts:8:30 - error TS2322: Type '{ a: number; b: number; }' is not assignable to type 'ITest'.
  Object literal may only specify known properties, and 'b' does not exist in type 'ITest'.

8     const c: ITest = { a: 3, b: 3 }
                               ~~~~

index.ts:11:20 - error TS2322: Type '{ a: number; b: number; }' is not assignable to type 'ITest'.
  Object literal may only specify known properties, and 'b' does not exist in type 'ITest'.

11     return { a: 3, b: 3 };
                      ~~~~

index.ts:14:20 - error TS2322: Type '{ a: number; b: number; }' is not assignable to type 'ITest'.
  Object literal may only specify known properties, and 'b' does not exist in type 'ITest'.

14     return { a: 3, b: 3 };
                      ~~~~

PS C:\Code\test> tslint -c tslint.json -p tsconfig.json index.ts

ERROR: C:/Code/test/index.ts[5, 12]: Type assertion on object literals is forbidden, use a type annotation instead.
```

Can you provide a test case of code where you would want an additional complaint from TSLint that doesn't already cause a complaint with `no-object-literal-type-assertion` + TypeScript?

**Comment by Z3roCoder** [comment]
Perhaps the behavior changed over the last two years since I posted the issue.
I'll try to reproduce it when I get to my Workstation and respond correspondingly

**Comment by JoshuaKGoldberg** [comment]
👋 @Z3roCoder I'm going to close this issue to keep the tracker clean, but please do post again if you have a repro! Sounds like this was fixed over time _(hurray!)_
